core/irCheck.ml

```diff
@@ -5,30 +5,85 @@ open Ir
-    | `Value v -> "\noccuring in value: " ^ Ir.string_of_value v
-    | `Special s -> "\noccuring in special tail computation: " ^ Ir.string_of_special s
-    | `Binding b -> "\noccuring in binding: " ^ Ir.string_of_binding b
-    | `None -> "" in
-  raise (IRTypeError (msg ^ occurrence_string))
-
-let handle_ir_type_error error alternative =
-  match error with
-    | IRTypeError msg
-    | TypeUtils.TypeDestructionError msg ->
-      if fail_on_ir_type_error then
-        failwith ("IR Type Error: " ^ msg)
-      else
+  let occurrence_string = string_of_occurrence occurrence in
+  raise (Errors.IRTypeError (msg ^ occurrence_string))
+
+
+(* Evalutes a lazy value and handles exceptions raised while doing so.
+   If we are supposed to fail on IR exceptions, we don't catch any of them.
+   If we are supposed to continue after IR type errors, we print a debug
+   message and return the alternative value in case of an exception. *)
+let handle_ir_type_error lazy_val alternative occurrence =
+  if fail_on_ir_type_error then
+    (* All exceptions are left unhandled, leaving them for the error handling
+       facilities outside of the IR type checker *)
+    Lazy.force lazy_val
+  else
+    (* Catch all errors, print them to debug output, but continue by
+       returning the alternative value *)
+    try
+      Lazy.force lazy_val
+    with
+      | Errors.IRTypeError msg  ->
+        (* We saw an IR error, but we are not supposed to abort.
+           Print the error and return the alternative value *)
         Debug.print
           ("\n--------------------------------------------------------------------\n" ^
-            "Continuing after IR Type Error: " ^
-            msg ^
-            "\n-------------------------------------------------------------------\n" ); alternative
-    |  e-> failwith ("Unknown exception during IR type-checking: " ^ Printexc.to_string e)
+           "Continuing after IR Type Error:\n" ^
+           msg ^
+           "\n-------------------------------------------------------------------\n" );
+        alternative
+      | exn ->
+        let msg =
+          "Encountered exception during IR type-checking, saying:\n" ^
+          Printexc.to_string exn in
+        let occurrence_string = string_of_occurrence occurrence in
+        Debug.print (msg ^ occurrence_string);
+        alternative
+
+
+
+
+(* Some of the helper functions we use raise their own exceptions.
+   We translate them to IR type exceptions here *)
+let translate_helper_exns (f : 'a -> 'b) (x : 'a) occurrence : 'b =
+  try
+    f x
+  with
+    | TypeUtils.TypeDestructionError msg ->
+      raise_ir_type_error msg occurrence
+    | Instantiate.ArityMismatch (takes, given) ->
+      let msg =
+        Printf.sprintf "Arity mismatch during type application/instantiation. \
+                        Providing %d type arguments to function that takes %d." given takes in
+      raise_ir_type_error msg occurrence
+
```

Comment:
Why so many empty lines?

Reply:
Indeed, here it doesn't make much sense, removed them